File: src/lc/p3/py/longestSubString.py
from typing import Dict

def lengthOfLongestSubstring(s: str) -> int:
    if len(s) == 0: 
        return 0
    
    str_len_longest: int = 1
    character_pos: Dict[str, int] = {s[0]:0}
    sub_str_start_pos = 0
    sub_str_len = 1

    for n in range(1, len(s)):
        c = s[n]
        if not c in character_pos:
            character_pos[c] = n
            sub_str_len += 1
            if sub_str_len > str_len_longest:
                str_len_longest = sub_str_len
        else:
            last_pos = character_pos[c]
            character_pos[c] = n
            
            if last_pos >= sub_str_start_pos:
                sub_str_start_pos = last_pos + 1
                sub_str_len = n - sub_str_start_pos + 1
            else:
                sub_str_len += 1
                if sub_str_len > str_len_longest:
                    str_len_longest = sub_str_len
            

    return str_len_longest

File: src/lc/p3/java/LongestSubString.java
import java.util.HashMap;

public class LongestSubString{
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0) return 0;
        
        Integer subStrLongestLength = 1;
        HashMap<Character, Integer> charPosMap = new HashMap<Character, Integer>();
        charPosMap.put(s.charAt(0), 0);
        Integer sub_str_start_pos = 0;
        Integer sub_str_len = 1;

        for(int i = 1; i < s.length(); ++i){
            Character c = s.charAt(i);
            if(charPosMap.containsKey(c)){
                Integer last_pos = charPosMap.get(c);
                if(last_pos >= sub_str_start_pos){
                    sub_str_start_pos = last_pos+1;
                    sub_str_len = i - sub_str_start_pos + 1;
                }else{
                    sub_str_len += 1;
                }
                charPosMap.put(c, i);
            }else{
                charPosMap.put(c, i);
                sub_str_len += 1;
            }

            if(sub_str_len > subStrLongestLength) subStrLongestLength = sub_str_len;
            //System.out.println("i=" + i + " sub_str_len: " + sub_str_len + " sub_str_start_pos: "+ sub_str_start_pos + "subStrLongestLength: "+ subStrLongestLength);
        }

        return subStrLongestLength;
    }
}